p2p/net/swarm/swarm_stream.go

```diff
@@ -75,9 +75,19 @@ func (s *Stream) Write(p []byte) (int, error) {
 	return n, err
 }
 
+type asyncCloser interface {
+	AsyncClose(onDone func()) error
+}
+
 // Close closes the stream, closing both ends and freeing all associated
 // resources.
 func (s *Stream) Close() error {
+	if as, ok := s.stream.(asyncCloser); ok {
+		err := as.AsyncClose(func() {
+			s.closeAndRemoveStream()
+		})
+		return err
+	}
```

Comment:
This means consumers can't acutally know if the remote ACKed received data (since `.Write` does not block for acks).
But I checked and we don't promise this in the first place:
> ```go
> // Close may be asynchronous and _does not_ guarantee receipt of the data.
> ```

I guess it's fine then.

Reply:
yeah, there is no guarantee that the data will be delivered after Close so the semantics seem okay.